File: ios/Classes/Data/MarkerStore.swift
import Foundation

final class MarkerStore {
  private var markers: [MarkerMessage] = []

  func set(markers: [MarkerMessage]) {
    self.markers = markers
      .filter(isRenderable)
      .sorted { $0.timestamp < $1.timestamp }
  }

  func add(marker: MarkerMessage) {
    guard isRenderable(marker) else {
      return
    }
    let insertIndex = lowerBound(for: marker.timestamp)
    markers.insert(marker, at: insertIndex)
  }

  func clear() {
    markers.removeAll()
  }

  func visibleMarkers(startTimestamp: Int64, endTimestamp: Int64) -> [MarkerMessage] {
    guard !markers.isEmpty else {
      return []
    }
    let startIndex = lowerBound(for: startTimestamp)
    var visible: [MarkerMessage] = []
    var index = startIndex
    while index < markers.count {
      let marker = markers[index]
      if marker.timestamp > endTimestamp {
        break
      }
      visible.append(marker)
      index += 1
    }
    return visible
  }

  private func lowerBound(for timestamp: Int64) -> Int {
    var low = 0
    var high = markers.count
    while low < high {
      let mid = (low + high) / 2
      if markers[mid].timestamp < timestamp {
        low = mid + 1
      } else {
        high = mid
      }
    }
    return low
  }

  private func isRenderable(_ marker: MarkerMessage) -> Bool {
    !marker.id.isEmpty &&
      marker.timestamp >= 0 &&
      marker.price.isFinite &&
      (marker.type == "buy" || marker.type == "sell")
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/data/MarkerStore.kt
package com.tradechart.plugin.data

import com.tradechart.plugin.bridge.generated.MarkerMessage

class MarkerStore {
    private val markers = mutableListOf<MarkerMessage>()

    fun set(items: List<MarkerMessage>) {
        markers.clear()
        markers.addAll(items.filter(::isRenderable))
        markers.sortBy { it.timestamp }
    }

    fun add(marker: MarkerMessage) {
        if (!isRenderable(marker)) {
            return
        }
        val insertIndex = markers.binarySearch { existing ->
            existing.timestamp.compareTo(marker.timestamp)
        }.let { index ->
            if (index >= 0) {
                index + 1
            } else {
                -index - 1
            }
        }
        markers.add(insertIndex, marker)
    }

    fun clear() {
        markers.clear()
    }

    fun visibleMarkers(
        startTimestamp: Long,
        endTimestamp: Long,
    ): List<MarkerMessage> {
        if (markers.isEmpty()) {
            return emptyList()
        }
        val startIndex = markers.binarySearch { marker ->
            marker.timestamp.compareTo(startTimestamp)
        }.let { index ->
            if (index >= 0) {
                index
            } else {
                -index - 1
            }
        }

        val visible = ArrayList<MarkerMessage>()
        for (index in startIndex until markers.size) {
            val marker = markers[index]
            if (marker.timestamp > endTimestamp) {
                break
            }
            visible.add(marker)
        }
        return visible
    }

    private fun isRenderable(marker: MarkerMessage): Boolean {
        return marker.id.isNotBlank() &&
            marker.timestamp >= 0 &&
            marker.price.isFinite() &&
            (marker.type == "buy" || marker.type == "sell")
    }
}
